[Grid] Improve Grid2 upgrade experience (#45304)
mui/material-ui: docs/data/material/migration/upgrade-to-grid-v2/upgrade-to-grid-v2.md

// The updated Grid component is named Unstable_Grid2
+import Grid from '@mui/material/Unstable_Grid2';
```

</codeblock>

### 2. Remove legacy props

The `item` and `zeroMinWidth` props have been removed in the updated Grid.
You can safely remove them:

```diff
-<Grid item zeroMinWidth>
+<Grid>
```

### 3. Update the size props

:::warning
Skip this step if you're using Material UI v5.
:::

In the legacy Grid, the size props were named to correspond with the theme's breakpoints.
For the default theme, these were: `xs`, `sm`, `md`, `lg`, `xl`

Starting from Material UI v6, these props are renamed to `size` on the updated Grid:

```diff
 <Grid
-  xs={12}
-  sm={6}
+  size={{ xs: 12, sm: 6 }}
 >
```

If the size is the same for all breakpoints, then you can use a single value:

```diff
-<Grid xs={6}>
+<Grid size={6}>
```

Additionally, the `true` value for the size props was renamed to `"grow"`:

```diff
-<Grid xs>
+<Grid size="grow">
```

You can use the following codemod to update the size props:

```bash v6
npx @mui/codemod@latest v6.0.0/grid-v2-props <path/to/folder>
```

:::warning
The codemod requires [updating the imports](#update-the-import) beforehand.
:::

> ### 4. Opt-in to legacy negative margins

Reviewer: In this context, "opt in" is a verb so it doesn't get hyphenated. The hyphen is only for when it's used as an adjective to describe a noun, as in "this is an opt-in feature."

```suggestion
### 4. Opt in to legacy negative margins
```

:::warning
Skip this step if you're using Material UI v6 or v7.
:::

If you're using Material UI v5 and want to apply the negative margins similar to the legacy Grid, specify `disableEqualOverflow={true}` on the grid container.
To apply to all grids, add the default props to the theme:

```js
import { createTheme, ThemeProvider } from '@mui/material/styles';
import Grid from '@mui/material/Unstable_Grid2';

const theme = createTheme({
  components: {
    MuiGrid2: {
      defaultProps: {
        // all grids under this theme will apply
        // negative margin on the top and left sides.
        disableEqualOverflow: true,
      },
    },
  },
});

function Demo() {
  return (
    <ThemeProvider theme={theme}>
      <Grid container>...grids</Grid>
    </ThemeProvider>
  );
}
```

## Common issues

> ### Direction column or column-reverse

Reviewer: I think plain English is preferable here over the CSS terminology since it appears grammatically incorrect without backticks.

```suggestion
### Column direction
```

> Using `direction="column"` or `direction="column-reverse"` is not supported on [the legacy Grid](/material-ui/react-grid/#direction-column-column-reverse) nor on [the updated Grid](/material-ui/react-grid2/#direction-column-or-column-reverse).

Reviewer: ```suggestion
Using `direction="column"` or `direction="column-reverse"` is not supported on [the legacy Grid](/material-ui/react-grid/#direction-column-column-reverse) nor on [the updated Grid2](/material-ui/react-grid2/#direction-column-or-column-reverse).
```

If your layout used the legacy Grid with these values, it might break when you switch to the updated Grid.

> If you need vertical layout, follow the instructions in the [updated Grid documentation](/material-ui/react-grid2/#direction-column-or-column-reverse).

Reviewer: ```suggestion
If you need a vertical layout, follow the instructions in the [Grid2 documentation](/material-ui/react-grid2/#direction-column-or-column-reverse).
```

## Documentation pages

> - Updated Grid:

Reviewer: ```suggestion
- Grid2:
```
